modules/transport-netty4/src/test/java/org/elasticsearch/http/netty4/Netty4HttpServerPipeliningTests.java

```diff
@@ -115,98 +112,44 @@ class CustomNettyHttpServerTransport extends Netty4HttpServerTransport {
             );
         }
 
-        @Override
-        public ChannelHandler configureServerChannelHandler() {
-            return new CustomHttpChannelHandler(this, executorService);
-        }
-
         @Override
         protected void doClose() {
             executorService.shutdown();
             super.doClose();
         }
 
-    }
-
-    private class CustomHttpChannelHandler extends Netty4HttpServerTransport.HttpChannelHandler {
-
-        private final ExecutorService executorService;
-
-        CustomHttpChannelHandler(Netty4HttpServerTransport transport, ExecutorService executorService) {
-            super(transport, transport.handlingSettings);
-            this.executorService = executorService;
-        }
-
         @Override
-        protected void initChannel(Channel ch) throws Exception {
-            super.initChannel(ch);
-            ch.pipeline().replace("handler", "handler", new PossiblySlowUpstreamHandler(executorService));
-        }
-
-    }
-
-    class PossiblySlowUpstreamHandler extends SimpleChannelInboundHandler<HttpPipelinedRequest> {
-
-        private final ExecutorService executorService;
-
-        PossiblySlowUpstreamHandler(ExecutorService executorService) {
-            this.executorService = executorService;
-        }
-
-        @Override
-        protected void channelRead0(ChannelHandlerContext ctx, HttpPipelinedRequest msg) throws Exception {
-            executorService.submit(new PossiblySlowRunnable(ctx, msg));
-        }
-
-        @Override
-        public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
-            logger.info("Caught exception", cause);
-            ctx.channel().close().sync();
-        }
-
-    }
-
-    class PossiblySlowRunnable implements Runnable {
-
-        private ChannelHandlerContext ctx;
-        private HttpPipelinedRequest pipelinedRequest;
-
-        PossiblySlowRunnable(ChannelHandlerContext ctx, HttpPipelinedRequest msg) {
-            this.ctx = ctx;
-            this.pipelinedRequest = msg;
-        }
-
-        @Override
-        public void run() {
-            try {
-                final String uri = pipelinedRequest.uri();
-
-                final ByteBuf buffer = Unpooled.copiedBuffer(uri, StandardCharsets.UTF_8);
-
-                HttpResponse response = pipelinedRequest.createResponse(
-                    RestStatus.OK,
-                    new BytesArray(uri.getBytes(StandardCharsets.UTF_8))
-                );
-                response.addHeader("content-length", Integer.toString(buffer.readableBytes()));
-
-                final boolean slow = uri.matches("/slow/\\d+");
-                if (slow) {
-                    try {
-                        Thread.sleep(scaledRandomIntBetween(500, 1000));
-                    } catch (InterruptedException e) {
-                        throw new RuntimeException(e);
+        public void incomingRequest(HttpRequest httpRequest, HttpChannel httpChannel) {
```

Comment:
This change puzzled me a bit in that you tried to keep most other things unchanged, but here went for a more radical change. I see the need for changes (using deleted types), but for my education perhaps you can explain the need for the radical change of the test.

Reply:
Sure :) See this change: https://github.com/elastic/elasticsearch/pull/86133/files#diff-5fb205d075055657d087e3445ff349bcfed01425594e604f7bf7b3d3a6dfefa9L325

we had the `pipelining` handler and then the `handler` as separate channel handlers before, now they're a single handler. So previously the test could just replace the "handler" with `PossiblySlowUpstreamHandler` and still test the logic in the unchanged "pipelining" handler. Now that both handlers are one, this wasn't possible any more.
So since I couldn't replace part of the handler, I left the pipeline as is and overrode the http server transport to catch the requests a little further down the stack.
Or put differently, we previously replaced this code https://github.com/elastic/elasticsearch/pull/86133/files#diff-a3368fd4145592d305c99a7c2b0a6c21c9bd6d45c828515aed1f228b69af2a79L29 in the test, now we run it in this test as well and overrode the code it calls.

Reply:
Thanks for the explanation!